{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rather than having to implement your own numerical recipes, Python has a rich array of helpful packages to use.\n",
    "\n",
    "#### Many of the most popular packages are supported by industry partners, so they have a strong following and are very likely to continue to improve and grow.\n",
    "\n",
    "These packages save you from having to reinvent the wheel each time, and allows you to lean on the experience of more experienced mathematicians, computer scientists, etc.\n",
    "\n",
    "Some of the most popular scientific Python packages include\n",
    "- numPy (basic numerical/matrix operations)\n",
    "- Pandas (manipulating tables)\n",
    "- SciPy (Scientific python-- numerical routines, machine learning, statistics)\n",
    "- matplotlib (plotting)\n",
    "\n",
    "Note that there are many more!  And like many things in programming/software, there are constant improvements and new projects to make many things easier/faster.\n",
    "\n",
    "To use most of them, you simply install the package with `pip install <package>`.  Once installed, you simply import:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note that I used an alias for the `numpy` package above.  I could have just written \"`import numpy`\" and left off the \"`as np`\" part.  However, it saves typing to use the alias.  For example,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 3])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = np.array([1,2,3])\n",
    "x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Otherwise, I would have to write `x = numpy.array([1,2,3])`.  Over time, that adds up."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python code is typically organized into packages (folders) and modules (python code files).\n",
    "\n",
    "#### Practically, this means that related methods and functionality are kept in the same folder.\n",
    "\n",
    "This is relevant because many scientific Python libraries are organized in this hierarchical manner and affects how the packages are loaded and used.  For example, the numPy library has a variety of methods for generating random numbers.  These reside in the `numpy.random` package.  If you dig through the numPy library, you will find a folder named \"random\" that contains a number of Python files, among others.  Deep down, a lot of numPy uses special code libraries written in high-performance languages like C++ and Fortran, so you may also see quite a few of those.     \n",
    "\n",
    "#### Jupyter notebooks are especially helpful because they can assist in finding things inside these packages without having to search online.  \n",
    "\n",
    "Given the size of these libraries, most of us only remember the most common names/methods and their usage; we certainly do not have everything memorized!  As an example, if you type `np.random.` (note the final dot) and hit your `<Tab>` key, a list of suggestions will appear.  Try it in the space below-- once you hit tab, a dropdown list should appear."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once you have the method of choice, using `<Shift>+<Tab>` can provide a description of the method and some help on using it.  For example, once you type `np.random.normal(` (note the opening parenthesis), you can hit `<Shift>+<Tab>` to show the help.  For the `np.random.normal` method below, it will tell you that it samples from a normal (Gaussian) distribution and tells you that the `loc` and `scale` parameters/arguments set the mean ($\\mu$) and standard deviation ($\\sigma$) (*not* variance) of the distribution.\n",
    "\n",
    "To get 10 random samples from a $\\mathcal{N}(\\mu=10, \\sigma=4)$ distribution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 4.78650282, 10.67362293, 13.65521029,  9.99666593,  8.55533101,\n",
       "       12.54949754, 13.90809602, 14.47902059, 16.33316439,  8.64775273])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.normal(loc=10, scale=4, size=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Suffice to say, there is a lot in these packages, and we cannot cover it in a short workshop.  \n",
    "\n",
    "**However**, we will focus the hands-on portions on (hopefully) useful libraries and techniques that can aid in getting everyone started.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### While numPy is very \"low-level\" and deals with a lot of numerical matrix manipulation, the Pandas package is more generally useful for researchers in life-sciences.\n",
    "\n",
    "Unless you are intimately familiar with matrix computations like QR-decomposition, SVD, etc. you will not likely use numPy directly.  Most users may only use the `numpy.random` package like shown above.  In our experience, it is more useful to explore the Pandas package.  \n",
    "\n",
    "I tend to treat Pandas as a non-visual Excel spreadsheet, as it operates on the same table-based data.  In fact, Pandas has special functions for importing/reading Excel files.\n",
    "\n",
    "Pandas tables are often called \"data frames\", following the name adopted in the R community.\n",
    "\n",
    "Let's import the package:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's read a gene expression matrix.\n",
    "\n",
    "This particular example is a cell-line experiment in a simple 3 vs. 3 setup comparing treated and control conditions.  Within each condition, the samples are *biological* replicates.\n",
    "\n",
    "We can import in a variety of ways, but Pandas provides a convenient `read_csv` function which allows us to read comma-separated (CSV) or tab-separated (TSV) files.  You just have to tell it which:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>SW1_Control</th>\n",
       "      <th>SW2_Control</th>\n",
       "      <th>SW3_Control</th>\n",
       "      <th>SW4_Treated</th>\n",
       "      <th>SW5_Treated</th>\n",
       "      <th>SW6_Treated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>45.77888</td>\n",
       "      <td>22.840605</td>\n",
       "      <td>42.016487</td>\n",
       "      <td>50.16839</td>\n",
       "      <td>29.507082</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  SW1_Control  SW2_Control  SW3_Control  SW4_Treated  \\\n",
       "0     DDX11L1      0.00000     0.000000     0.000000      0.00000   \n",
       "1      WASH7P     45.77888    22.840605    42.016487     50.16839   \n",
       "2  MIR1302-10      0.00000     0.000000     0.000000      0.00000   \n",
       "3     FAM138A      0.00000     0.000000     0.000000      0.00000   \n",
       "4      OR4G4P      0.00000     0.000000     0.000000      0.00000   \n",
       "\n",
       "   SW5_Treated  SW6_Treated  \n",
       "0     0.000000     0.000000  \n",
       "1    29.507082    23.742158  \n",
       "2     0.000000     0.000000  \n",
       "3     0.000000     0.000000  \n",
       "4     0.000000     0.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df1 = pd.read_csv('gene_expression.tsv', sep='\\t')\n",
    "expression_df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that Pandas automatically reads the column names from the file, and \"indexes\" each row by consecutive integers 0,1,...  (see the bold numbers on the left-hand side)\n",
    "\n",
    "Note that we told `read_csv` that the file had the columns separated by `<Tab>` characters by including the `sep=\"\\t\"` argument.  The string `\"\\t\"` means a tab-character.  This is consistent across computing languages and platforms.\n",
    "\n",
    "Depending on your preferences, you might instead choose to name/index each row by the gene instead of an integer.  In that case, you tell Pandas that you want the first (0- Python starts at zero!) column to be used as the index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SW1_Control</th>\n",
       "      <th>SW2_Control</th>\n",
       "      <th>SW3_Control</th>\n",
       "      <th>SW4_Treated</th>\n",
       "      <th>SW5_Treated</th>\n",
       "      <th>SW6_Treated</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>45.77888</td>\n",
       "      <td>22.840605</td>\n",
       "      <td>42.016487</td>\n",
       "      <td>50.16839</td>\n",
       "      <td>29.507082</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            SW1_Control  SW2_Control  SW3_Control  SW4_Treated  SW5_Treated  \\\n",
       "gene                                                                          \n",
       "DDX11L1         0.00000     0.000000     0.000000      0.00000     0.000000   \n",
       "WASH7P         45.77888    22.840605    42.016487     50.16839    29.507082   \n",
       "MIR1302-10      0.00000     0.000000     0.000000      0.00000     0.000000   \n",
       "FAM138A         0.00000     0.000000     0.000000      0.00000     0.000000   \n",
       "OR4G4P          0.00000     0.000000     0.000000      0.00000     0.000000   \n",
       "\n",
       "            SW6_Treated  \n",
       "gene                     \n",
       "DDX11L1        0.000000  \n",
       "WASH7P        23.742158  \n",
       "MIR1302-10     0.000000  \n",
       "FAM138A        0.000000  \n",
       "OR4G4P         0.000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df2 = pd.read_csv('gene_expression.tsv', sep='\\t', index_col=0)\n",
    "expression_df2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### By naming the rows by their gene, it helps when selecting and filtering data, as you will see below.\n",
    "\n",
    "#### Note that if your input file did **not** have column headers, you have to tell Pandas-- otherwise it might do something you do not expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0.1</th>\n",
       "      <th>0.2</th>\n",
       "      <th>0.3</th>\n",
       "      <th>0.4</th>\n",
       "      <th>0.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>45.77888</td>\n",
       "      <td>22.840605</td>\n",
       "      <td>42.016487</td>\n",
       "      <td>50.16839</td>\n",
       "      <td>29.507082</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G11P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.855524</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0        0.1        0.2       0.3        0.4        0.5\n",
       "DDX11L1                                                                     \n",
       "WASH7P        45.77888  22.840605  42.016487  50.16839  29.507082  23.742158\n",
       "MIR1302-10     0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "FAM138A        0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G4P         0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G11P        0.00000   0.000000   0.000000   0.00000   2.855524   0.000000\n",
       "OR4F5          0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.7   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.8   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('table_with_no_header.tsv', sep='\\t', index_col=0)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above, it took the first row (which was expressions for gene DDX11L1) and somehow \"coerced\" it into unique column names.  NOT what we wanted!\n",
    "\n",
    "Try again, this time telling Pandas there is no header by using the `header=None` argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>45.77888</td>\n",
       "      <td>22.840605</td>\n",
       "      <td>42.016487</td>\n",
       "      <td>50.16839</td>\n",
       "      <td>29.507082</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G11P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.855524</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     1          2          3         4          5          6\n",
       "0                                                                           \n",
       "DDX11L1        0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "WASH7P        45.77888  22.840605  42.016487  50.16839  29.507082  23.742158\n",
       "MIR1302-10     0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "FAM138A        0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G4P         0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G11P        0.00000   0.000000   0.000000   0.00000   2.855524   0.000000\n",
       "OR4F5          0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.7   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.8   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('table_with_no_header.tsv', sep='\\t', index_col=0, header=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, that's better, but the column names are not particularly meaningful.  Let's assume we knew the first three were control and the last three were treated.  We can tell Pandas these names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C1</th>\n",
       "      <th>C2</th>\n",
       "      <th>C3</th>\n",
       "      <th>T1</th>\n",
       "      <th>T2</th>\n",
       "      <th>T3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>45.77888</td>\n",
       "      <td>22.840605</td>\n",
       "      <td>42.016487</td>\n",
       "      <td>50.16839</td>\n",
       "      <td>29.507082</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G11P</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.855524</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    C1         C2         C3        T1         T2         T3\n",
       "DDX11L1        0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "WASH7P        45.77888  22.840605  42.016487  50.16839  29.507082  23.742158\n",
       "MIR1302-10     0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "FAM138A        0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G4P         0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "OR4G11P        0.00000   0.000000   0.000000   0.00000   2.855524   0.000000\n",
       "OR4F5          0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.7   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000\n",
       "RP11-34P13.8   0.00000   0.000000   0.000000   0.00000   0.000000   0.000000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = ['C1', 'C2', 'C3', 'T1', 'T2', 'T3']\n",
    "df = pd.read_csv('table_with_no_header.tsv', sep='\\t', index_col=0, names=column_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, assume you forgot the `sep='\\t'` argument.  What happens?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C2</th>\n",
       "      <th>C3</th>\n",
       "      <th>T1</th>\n",
       "      <th>T2</th>\n",
       "      <th>T3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P\\t45.7788798161081\\t22.840605069235\\t42.016486571677\\t50.168389863934\\t29.5070817011913\\t23.7421576952357</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G11P\\t0\\t0\\t0\\t0\\t2.85552403559915\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8\\t0\\t0\\t0\\t0\\t0\\t0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    C2  C3  T1  T2  T3\n",
       "C1                                                                    \n",
       "DDX11L1\\t0\\t0\\t0\\t0\\t0\\t0                          NaN NaN NaN NaN NaN\n",
       "WASH7P\\t45.7788798161081\\t22.840605069235\\t42.0... NaN NaN NaN NaN NaN\n",
       "MIR1302-10\\t0\\t0\\t0\\t0\\t0\\t0                       NaN NaN NaN NaN NaN\n",
       "FAM138A\\t0\\t0\\t0\\t0\\t0\\t0                          NaN NaN NaN NaN NaN\n",
       "OR4G4P\\t0\\t0\\t0\\t0\\t0\\t0                           NaN NaN NaN NaN NaN\n",
       "OR4G11P\\t0\\t0\\t0\\t0\\t2.85552403559915\\t0           NaN NaN NaN NaN NaN\n",
       "OR4F5\\t0\\t0\\t0\\t0\\t0\\t0                            NaN NaN NaN NaN NaN\n",
       "RP11-34P13.7\\t0\\t0\\t0\\t0\\t0\\t0                     NaN NaN NaN NaN NaN\n",
       "RP11-34P13.8\\t0\\t0\\t0\\t0\\t0\\t0                     NaN NaN NaN NaN NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('table_with_no_header.tsv', index_col=0, names=column_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whoa!  By default, `read_csv` is expecting comma-separated values (hence the name).  It did not find any commas, so it simply read the ENTIRE line as a single item.  Since we gave it 6 column names, it filled in six columns of \"junk\".  `NaN` means \"Not a Number\" and is essentially a placeholder for data that is missing.\n",
    "\n",
    "The point of the above \"mistakes\" is to show that there any many different ways to adjust how the `read_csv` function behaves.  Depending on the files you are given or use, you may have to employ various combinations of the arguments shown above to get your tables just right.  If you forget, you can either search online, or use the `<Shift>+<Tab>` shortcut described above.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We will now also load a table of differential expression results.\n",
    "\n",
    "This table was prepared by DESeq2 software (a differential expression analysis package in the R language).\n",
    "\n",
    "This time, we know the columns have names and are separated by commas, so we do *not* need to use the `names` or `sep=\"\\t\"` arguments.  We note that the first column, \"gene\", is a good candidate for the \"index\" of the rows. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>foldChange</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>35.6756</td>\n",
       "      <td>36.878657</td>\n",
       "      <td>34.472543</td>\n",
       "      <td>0.934756</td>\n",
       "      <td>-0.097338</td>\n",
       "      <td>0.876438</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            baseMean          C          T  foldChange  log2FoldChange  \\\n",
       "gene                                                                     \n",
       "DDX11L1       0.0000   0.000000   0.000000         NaN             NaN   \n",
       "WASH7P       35.6756  36.878657  34.472543    0.934756       -0.097338   \n",
       "MIR1302-10    0.0000   0.000000   0.000000         NaN             NaN   \n",
       "FAM138A       0.0000   0.000000   0.000000         NaN             NaN   \n",
       "OR4G4P        0.0000   0.000000   0.000000         NaN             NaN   \n",
       "\n",
       "                pval  padj  \n",
       "gene                        \n",
       "DDX11L1          NaN   NaN  \n",
       "WASH7P      0.876438   1.0  \n",
       "MIR1302-10       NaN   NaN  \n",
       "FAM138A          NaN   NaN  \n",
       "OR4G4P           NaN   NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge_results = pd.read_csv('differential_results.csv', index_col='gene')\n",
    "dge_results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We can perform many \"standard\" operations on these tables, such as sorting.  Below, we sort by the raw p-value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>foldChange</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDKN1A</th>\n",
       "      <td>16888.298028</td>\n",
       "      <td>8199.879130</td>\n",
       "      <td>25576.716926</td>\n",
       "      <td>3.119158</td>\n",
       "      <td>1.641157</td>\n",
       "      <td>1.505242e-180</td>\n",
       "      <td>4.876682e-176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-572C15.6</th>\n",
       "      <td>5798.732589</td>\n",
       "      <td>2607.993295</td>\n",
       "      <td>8989.471882</td>\n",
       "      <td>3.446892</td>\n",
       "      <td>1.785296</td>\n",
       "      <td>6.031485e-175</td>\n",
       "      <td>9.770402e-171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP3</th>\n",
       "      <td>13594.636434</td>\n",
       "      <td>7170.832682</td>\n",
       "      <td>20018.440186</td>\n",
       "      <td>2.791648</td>\n",
       "      <td>1.481117</td>\n",
       "      <td>1.231601e-145</td>\n",
       "      <td>1.330047e-141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPRC5A</th>\n",
       "      <td>30020.288348</td>\n",
       "      <td>16267.924899</td>\n",
       "      <td>43772.651797</td>\n",
       "      <td>2.690734</td>\n",
       "      <td>1.428000</td>\n",
       "      <td>2.028961e-141</td>\n",
       "      <td>1.380890e-137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACNA2D1</th>\n",
       "      <td>19970.721096</td>\n",
       "      <td>10848.555878</td>\n",
       "      <td>29092.886314</td>\n",
       "      <td>2.681729</td>\n",
       "      <td>1.423163</td>\n",
       "      <td>2.131135e-141</td>\n",
       "      <td>1.380890e-137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   baseMean             C             T  foldChange  \\\n",
       "gene                                                                  \n",
       "CDKN1A         16888.298028   8199.879130  25576.716926    3.119158   \n",
       "RP11-572C15.6   5798.732589   2607.993295   8989.471882    3.446892   \n",
       "IGFBP3         13594.636434   7170.832682  20018.440186    2.791648   \n",
       "GPRC5A         30020.288348  16267.924899  43772.651797    2.690734   \n",
       "CACNA2D1       19970.721096  10848.555878  29092.886314    2.681729   \n",
       "\n",
       "               log2FoldChange           pval           padj  \n",
       "gene                                                         \n",
       "CDKN1A               1.641157  1.505242e-180  4.876682e-176  \n",
       "RP11-572C15.6        1.785296  6.031485e-175  9.770402e-171  \n",
       "IGFBP3               1.481117  1.231601e-145  1.330047e-141  \n",
       "GPRC5A               1.428000  2.028961e-141  1.380890e-137  \n",
       "CACNA2D1             1.423163  2.131135e-141  1.380890e-137  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge_results_sorted = dge_results.sort_values('pval')\n",
    "dge_results_sorted.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We will now briefly discuss how to filter the rows of these data frames.  \n",
    "\n",
    "#### We need to first discuss how to \"select\" rows and columns.  We do this by addressing the row and column we want.  Since we named the rows by the name, this is very easy.  For example, to get the adjusted p-value of PPARG we write:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0143551492563152"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# note that we select by [row, column]\n",
    "dge_results.loc['PPARG', 'padj']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The use of `loc` can be read as \"give me the data in the following \"location\".  We tell pandas we want the data at the location of the \"PPARG\" row and \"padj\" column.\n",
    "\n",
    "#### Note that if we did not name the rows by the gene names, this would be a lot more difficult.  We would have to know the row number, which is completely dependent on the order of the rows.  What if someone sorted by p-value?  Or by gene name?  By labeling the rows with the gene name, you are making your code \"flexible\".  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Above we selected both the row and the column.  If we leave out 'padj', we get all the information about the PPARG row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "baseMean          863.294053\n",
       "C                 956.831931\n",
       "T                 769.756174\n",
       "foldChange          0.804484\n",
       "log2FoldChange     -0.313864\n",
       "pval                0.001249\n",
       "padj                0.014355\n",
       "Name: PPARG, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge_results.loc['PPARG']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We can also select all the values in a specific column.  Note, however, that now we do NOT use the `loc` for columns.  If we give pandas only one \"location\", it defaults to search through the rows, as shown above for \"PPARG\".  If we leave it out, as shown below, it looks through the columns.\n",
    "\n",
    "Note that this is a lot of data (>50,000 rows), and the notebook will automatically show only the first and last portions.  Otherwise your screen would quickly fill up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "DDX11L1                   NaN\n",
       "WASH7P           1.000000e+00\n",
       "MIR1302-10                NaN\n",
       "FAM138A                   NaN\n",
       "OR4G4P                    NaN\n",
       "OR4G11P          1.000000e+00\n",
       "OR4F5                     NaN\n",
       "RP11-34P13.7              NaN\n",
       "RP11-34P13.8              NaN\n",
       "CICP27                    NaN\n",
       "AL627309.1       1.000000e+00\n",
       "RP11-34P13.15             NaN\n",
       "RP11-34P13.16             NaN\n",
       "RP11-34P13.14    1.000000e+00\n",
       "RP11-34P13.13    1.000000e+00\n",
       "RNU6-1100P                NaN\n",
       "RP11-34P13.9              NaN\n",
       "AP006222.2       1.000000e+00\n",
       "AP006222.1                NaN\n",
       "RP4-669L17.10    1.000000e+00\n",
       "RP4-669L17.8              NaN\n",
       "CICP7                     NaN\n",
       "RP4-669L17.4              NaN\n",
       "OR4F29                    NaN\n",
       "WBP1LP7                   NaN\n",
       "AL732372.1                NaN\n",
       "RP4-669L17.2              NaN\n",
       "RP5-857K21.15             NaN\n",
       "RP4-669L17.1              NaN\n",
       "RP5-857K21.1     1.000000e+00\n",
       "                     ...     \n",
       "MT-TQ                     NaN\n",
       "MT-TM            1.000000e+00\n",
       "MT-ND2           9.325576e-01\n",
       "MT-TW            1.000000e+00\n",
       "MT-TA                     NaN\n",
       "MT-TN                     NaN\n",
       "MT-TC            2.819543e-15\n",
       "MT-TY            6.690900e-07\n",
       "MT-CO1           8.968042e-12\n",
       "MT-TS1                    NaN\n",
       "MT-TD                     NaN\n",
       "MT-CO2           1.296667e-08\n",
       "MT-TK                     NaN\n",
       "MT-ATP8          1.000000e+00\n",
       "MT-ATP6          2.622271e-01\n",
       "MT-CO3           6.937202e-06\n",
       "MT-TG                     NaN\n",
       "MT-ND3           1.521772e-01\n",
       "MT-TR                     NaN\n",
       "MT-ND4L          3.069162e-01\n",
       "MT-ND4           6.343151e-06\n",
       "MT-TH            1.000000e+00\n",
       "MT-TS2           1.000000e+00\n",
       "MT-TL2                    NaN\n",
       "MT-ND5           3.160187e-03\n",
       "MT-ND6           4.089413e-08\n",
       "MT-TE            8.323890e-01\n",
       "MT-CYB           7.737368e-05\n",
       "MT-TT                     NaN\n",
       "MT-TP            2.289117e-04\n",
       "Name: padj, Length: 56638, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge_results['padj']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If we select something that does not exist, Pandas gives us an error:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'the label [JUNK] is not in the [index]'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1789\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1790\u001b[0;31m                     \u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1791\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36merror\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1784\u001b[0m                                .format(key=key,\n\u001b[0;32m-> 1785\u001b[0;31m                                        axis=self.obj._get_axis_name(axis)))\n\u001b[0m\u001b[1;32m   1786\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'the label [JUNK] is not in the [index]'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-43a9c05cb73a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdge_results\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'JUNK'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1476\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1477\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_apply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1478\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1479\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1480\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1909\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1910\u001b[0m         \u001b[0;31m# fall thru to straight lookup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1911\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1912\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1913\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1796\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1797\u001b[0m             \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1798\u001b[0;31m                 \u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1799\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1800\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/python_venvs/py36_venv/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36merror\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1783\u001b[0m                 raise KeyError(u\"the label [{key}] is not in the [{axis}]\"\n\u001b[1;32m   1784\u001b[0m                                .format(key=key,\n\u001b[0;32m-> 1785\u001b[0;31m                                        axis=self.obj._get_axis_name(axis)))\n\u001b[0m\u001b[1;32m   1786\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1787\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'the label [JUNK] is not in the [index]'"
     ]
    }
   ],
   "source": [
    "dge_results.loc['JUNK']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### While the error message can be a bit dense, it will typically tell you the reason for the error.  In this case, you are told that \"JUNK\" is not in the index (AKA the list of genes).  \n",
    "\n",
    "### In addition to selecting rows, columns, or \"cells\", you can use logic operations to create \"filters\" for rows and columns.\n",
    "\n",
    "Let's say we want to find genes which are significantly upregulated with a log-fold change of > 3 and an adjusted p-values (FDR) < 0.01.\n",
    "\n",
    "To do that, we can run \"tests\" to keep or discard rows and/or columns of our table.  For our application, we want to keep only those rows that have have BOTH a log-fold change >3 AND FDR < 0.01.  We can separately test whether the log2 fold-change is greater than 3 and the FDR is less than 0.01.\n",
    "\n",
    "In the figure below, we have our differential expression table on the left.  In the table on the right-side, we show the results of our \"tests\".  We align the two tables to show the correspondance of the rows in each table.  In the first column, we see that only two genes (g2 and g3) have log-fold change > 3.  In the second column, we see that g1, g3, and g4 have padj < 0.01.  Together, the only row that passes both \"tests\" is for g3 (green highlight) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![](filtering.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create the table on the right, we note that the syntax is very simple and intuitive.  Pandas is smart enough to \"expand\" the operation and apply it to the entire column.  For instance,\n",
    "\n",
    "```\n",
    "dge_results['log2FoldChange']\n",
    "```\n",
    "selects the `log2FoldChange` column.  We then want to check *each row* to see whether the value is > 3.  If you execute \n",
    "\n",
    "```\n",
    "dge_results['log2FoldChange'] > 3\n",
    "```\n",
    "\n",
    "You will see that it is a long array of True or False.  Pandas automatically applied the test to the entire column.\n",
    "\n",
    "#### We can thus create the right-hand table with the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "upregulated = dge_results['log2FoldChange'] > 3\n",
    "significant = dge_results['padj'] < 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Since we want rows/genes that pass BOTH conditions, we are looking for situations where we have True and True, as shown above in the green row.\n",
    "\n",
    "Python comes with logical AND (`&`), and Pandas is smart enough to infer that we want to run this comparison on each row separately:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# By combining with \"&\", we require that we satisfy BOTH conditions\n",
    "row_filter = upregulated & significant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finally, we use `row_filter` to filter the table.  Recall that if we are working with rows, we have to use the `loc` word.  \n",
    "\n",
    "Earlier, we used the gene names to filter the rows, but we can also use True or False values.  It keeps the rows where it is True and removes those that are False. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>foldChange</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TNFSF18</th>\n",
       "      <td>23.555357</td>\n",
       "      <td>5.084786</td>\n",
       "      <td>42.025928</td>\n",
       "      <td>8.265034</td>\n",
       "      <td>3.047021</td>\n",
       "      <td>2.720555e-07</td>\n",
       "      <td>7.983744e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-46A10.2</th>\n",
       "      <td>35.279041</td>\n",
       "      <td>7.690233</td>\n",
       "      <td>62.867849</td>\n",
       "      <td>8.175025</td>\n",
       "      <td>3.031223</td>\n",
       "      <td>3.421609e-08</td>\n",
       "      <td>1.196143e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-131K19.6</th>\n",
       "      <td>9.705560</td>\n",
       "      <td>1.698615</td>\n",
       "      <td>17.712504</td>\n",
       "      <td>10.427613</td>\n",
       "      <td>3.382337</td>\n",
       "      <td>5.067686e-04</td>\n",
       "      <td>6.706818e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPH</th>\n",
       "      <td>106.078733</td>\n",
       "      <td>18.553392</td>\n",
       "      <td>193.604074</td>\n",
       "      <td>10.434969</td>\n",
       "      <td>3.383354</td>\n",
       "      <td>3.753955e-32</td>\n",
       "      <td>2.027011e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPATA18</th>\n",
       "      <td>15.795512</td>\n",
       "      <td>2.369358</td>\n",
       "      <td>29.221667</td>\n",
       "      <td>12.333158</td>\n",
       "      <td>3.624470</td>\n",
       "      <td>1.965324e-06</td>\n",
       "      <td>4.860501e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ART3</th>\n",
       "      <td>4.886862</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.773723</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>7.554618e-04</td>\n",
       "      <td>9.428140e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYOZ2</th>\n",
       "      <td>10.087959</td>\n",
       "      <td>1.860177</td>\n",
       "      <td>18.315740</td>\n",
       "      <td>9.846233</td>\n",
       "      <td>3.299572</td>\n",
       "      <td>3.384259e-04</td>\n",
       "      <td>4.746459e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSG2</th>\n",
       "      <td>6.477009</td>\n",
       "      <td>0.324673</td>\n",
       "      <td>12.629345</td>\n",
       "      <td>38.898671</td>\n",
       "      <td>5.281649</td>\n",
       "      <td>4.113713e-04</td>\n",
       "      <td>5.583414e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUC19</th>\n",
       "      <td>10.371199</td>\n",
       "      <td>1.362882</td>\n",
       "      <td>19.379515</td>\n",
       "      <td>14.219508</td>\n",
       "      <td>3.829800</td>\n",
       "      <td>8.307923e-05</td>\n",
       "      <td>1.392447e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POSTN</th>\n",
       "      <td>19.779884</td>\n",
       "      <td>3.158869</td>\n",
       "      <td>36.400899</td>\n",
       "      <td>11.523396</td>\n",
       "      <td>3.526494</td>\n",
       "      <td>3.815001e-04</td>\n",
       "      <td>5.230571e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 baseMean          C           T  foldChange  log2FoldChange  \\\n",
       "gene                                                                           \n",
       "TNFSF18         23.555357   5.084786   42.025928    8.265034        3.047021   \n",
       "RP11-46A10.2    35.279041   7.690233   62.867849    8.175025        3.031223   \n",
       "RP13-131K19.6    9.705560   1.698615   17.712504   10.427613        3.382337   \n",
       "LIPH           106.078733  18.553392  193.604074   10.434969        3.383354   \n",
       "SPATA18         15.795512   2.369358   29.221667   12.333158        3.624470   \n",
       "ART3             4.886862   0.000000    9.773723         inf             inf   \n",
       "MYOZ2           10.087959   1.860177   18.315740    9.846233        3.299572   \n",
       "NSG2             6.477009   0.324673   12.629345   38.898671        5.281649   \n",
       "MUC19           10.371199   1.362882   19.379515   14.219508        3.829800   \n",
       "POSTN           19.779884   3.158869   36.400899   11.523396        3.526494   \n",
       "\n",
       "                       pval          padj  \n",
       "gene                                       \n",
       "TNFSF18        2.720555e-07  7.983744e-06  \n",
       "RP11-46A10.2   3.421609e-08  1.196143e-06  \n",
       "RP13-131K19.6  5.067686e-04  6.706818e-03  \n",
       "LIPH           3.753955e-32  2.027011e-29  \n",
       "SPATA18        1.965324e-06  4.860501e-05  \n",
       "ART3           7.554618e-04  9.428140e-03  \n",
       "MYOZ2          3.384259e-04  4.746459e-03  \n",
       "NSG2           4.113713e-04  5.583414e-03  \n",
       "MUC19          8.307923e-05  1.392447e-03  \n",
       "POSTN          3.815001e-04  5.230571e-03  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significantly_upregulated = dge_results.loc[row_filter]\n",
    "significantly_upregulated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now, using those significantly upregulated genes, what are the observed expressions?  So far, we only have the DESEq2 results.\n",
    "\n",
    "We can get the genes by selecting the index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TNFSF18', 'RP11-46A10.2', 'RP13-131K19.6', 'LIPH', 'SPATA18', 'ART3',\n",
       "       'MYOZ2', 'NSG2', 'MUC19', 'POSTN'],\n",
       "      dtype='object', name='gene')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significantly_upregulated.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now use that to filter the expression matrix.  Again, since we are filtering rows, we have to use `loc`.  As you see in the cell above, we have a list of gene names, so the filtering is very natural:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SW1_Control</th>\n",
       "      <th>SW2_Control</th>\n",
       "      <th>SW3_Control</th>\n",
       "      <th>SW4_Treated</th>\n",
       "      <th>SW5_Treated</th>\n",
       "      <th>SW6_Treated</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TNFSF18</th>\n",
       "      <td>7.792150</td>\n",
       "      <td>5.191047</td>\n",
       "      <td>2.271161</td>\n",
       "      <td>41.209749</td>\n",
       "      <td>30.458923</td>\n",
       "      <td>54.409111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-46A10.2</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>13.496721</td>\n",
       "      <td>5.677904</td>\n",
       "      <td>49.272526</td>\n",
       "      <td>54.254957</td>\n",
       "      <td>85.076065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-131K19.6</th>\n",
       "      <td>2.922056</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>1.135581</td>\n",
       "      <td>17.021418</td>\n",
       "      <td>12.373937</td>\n",
       "      <td>23.742158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPH</th>\n",
       "      <td>25.324487</td>\n",
       "      <td>15.573140</td>\n",
       "      <td>14.762549</td>\n",
       "      <td>168.422452</td>\n",
       "      <td>204.645889</td>\n",
       "      <td>207.743880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPATA18</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>2.076419</td>\n",
       "      <td>1.135581</td>\n",
       "      <td>40.313885</td>\n",
       "      <td>28.555240</td>\n",
       "      <td>18.795875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ART3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.166913</td>\n",
       "      <td>15.229462</td>\n",
       "      <td>6.924796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYOZ2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>4.542323</td>\n",
       "      <td>14.333826</td>\n",
       "      <td>23.796034</td>\n",
       "      <td>16.817362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSG2</th>\n",
       "      <td>0.974019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.958641</td>\n",
       "      <td>19.036827</td>\n",
       "      <td>9.892566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUC19</th>\n",
       "      <td>0.974019</td>\n",
       "      <td>3.114628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.396603</td>\n",
       "      <td>21.892351</td>\n",
       "      <td>13.849592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POSTN</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>4.542323</td>\n",
       "      <td>31.355244</td>\n",
       "      <td>58.062322</td>\n",
       "      <td>19.785131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               SW1_Control  SW2_Control  SW3_Control  SW4_Treated  \\\n",
       "gene                                                                \n",
       "TNFSF18           7.792150     5.191047     2.271161    41.209749   \n",
       "RP11-46A10.2      3.896075    13.496721     5.677904    49.272526   \n",
       "RP13-131K19.6     2.922056     1.038209     1.135581    17.021418   \n",
       "LIPH             25.324487    15.573140    14.762549   168.422452   \n",
       "SPATA18           3.896075     2.076419     1.135581    40.313885   \n",
       "ART3              0.000000     0.000000     0.000000     7.166913   \n",
       "MYOZ2             0.000000     1.038209     4.542323    14.333826   \n",
       "NSG2              0.974019     0.000000     0.000000     8.958641   \n",
       "MUC19             0.974019     3.114628     0.000000    22.396603   \n",
       "POSTN             3.896075     1.038209     4.542323    31.355244   \n",
       "\n",
       "               SW5_Treated  SW6_Treated  \n",
       "gene                                     \n",
       "TNFSF18          30.458923    54.409111  \n",
       "RP11-46A10.2     54.254957    85.076065  \n",
       "RP13-131K19.6    12.373937    23.742158  \n",
       "LIPH            204.645889   207.743880  \n",
       "SPATA18          28.555240    18.795875  \n",
       "ART3             15.229462     6.924796  \n",
       "MYOZ2            23.796034    16.817362  \n",
       "NSG2             19.036827     9.892566  \n",
       "MUC19            21.892351    13.849592  \n",
       "POSTN            58.062322    19.785131  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_subset = expression_df2.loc[significantly_upregulated.index]\n",
    "expression_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### As usual, we can filter both the rows AND the columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SW1_Control</th>\n",
       "      <th>SW2_Control</th>\n",
       "      <th>SW3_Control</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TNFSF18</th>\n",
       "      <td>7.792150</td>\n",
       "      <td>5.191047</td>\n",
       "      <td>2.271161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-46A10.2</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>13.496721</td>\n",
       "      <td>5.677904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-131K19.6</th>\n",
       "      <td>2.922056</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>1.135581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPH</th>\n",
       "      <td>25.324487</td>\n",
       "      <td>15.573140</td>\n",
       "      <td>14.762549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPATA18</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>2.076419</td>\n",
       "      <td>1.135581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ART3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYOZ2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>4.542323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NSG2</th>\n",
       "      <td>0.974019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUC19</th>\n",
       "      <td>0.974019</td>\n",
       "      <td>3.114628</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POSTN</th>\n",
       "      <td>3.896075</td>\n",
       "      <td>1.038209</td>\n",
       "      <td>4.542323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               SW1_Control  SW2_Control  SW3_Control\n",
       "gene                                                \n",
       "TNFSF18           7.792150     5.191047     2.271161\n",
       "RP11-46A10.2      3.896075    13.496721     5.677904\n",
       "RP13-131K19.6     2.922056     1.038209     1.135581\n",
       "LIPH             25.324487    15.573140    14.762549\n",
       "SPATA18           3.896075     2.076419     1.135581\n",
       "ART3              0.000000     0.000000     0.000000\n",
       "MYOZ2             0.000000     1.038209     4.542323\n",
       "NSG2              0.974019     0.000000     0.000000\n",
       "MUC19             0.974019     3.114628     0.000000\n",
       "POSTN             3.896075     1.038209     4.542323"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_samples = [x for x in expression_df2.columns if x.endswith('_Control')]\n",
    "expression_subset_control = expression_df2.loc[significantly_upregulated.index, control_samples]\n",
    "expression_subset_control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now that we know some basics in data loading/manipulation/filtering, we can begin to think about analyses and data visualizations, which we cover in the next notebook "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
