---
title: "RNASeq Data Exploratory Basics"
author: "Yaoyu E. Wang"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

# RNASeq Data Exploratory Analysis
We first load the RNA-Seq count matrix from 'RNASeqData.Rdata'.  This is a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count matrix by gene.    

### Key commands for this sections:

function| package | description
|---|---|---|
subset()|base| Return subsets of vectors, matrices or data frames
max()|base|Returns the (regular or parallel) maxima in vector x
min()|base|Returns the (regular or parallel) minima in vector x
sd()|stats|This function computes the standard deviation of the values in vector x
rowMeans()|base|Form row and column sums and means for numeric arrays (or data frames)
apply()|base|Returns a vector or array or list of values obtained by applying a function to margins of an array or matrix
function()|base|the base mechanisms for defining new functions in the R language.
dist()|stats|computes and returns the distance matrix
hclust()|stats| Hierarchical cluster analysis on distance matrix
princomp()|stats| performs a principal components analysis on the given numeric data matrix and returns the results as an object of class princomp




## Explore RNASeq Count Data

### Load count and annotation files
Let's start with reading in our demo_raw_counts.tsv file.  It is a real 3 control v 3 treated bulk RNASeq data set, aligned by STAR and quantified using featureCount for gene counts.  The file is not normalized.  Since is a count file,  we can specify the **Gene** column is to be used as the row.names to obtain a numeric count matrix.

```{r load expr, echo = TRUE, include=TRUE}
# Set working directory to PROJECT_DIR
count_data=read.table("data/demo_raw_counts.tsv", 
                      header=TRUE, row.names='Gene')
head(count_data)           # The head(count_data) provides the first 6 row of the 'count_data'
```

Then, load in *sample_annotation.txt* for sample annotation.  The annotation file contains two columns: *sample*, *group* to indicate sample names and the group they belong to.  The data is loaded as a data.frame with two columns, and we can use **subset()** function to separate out samples groups and store in list. It takes form of:
```
result_subset = subset(input data.frame, selection criteria, select=c(header))

# if select is null, all columns are retained
```

For example, we can select out *control* and *treated* group with the following:
```{r load annotation, echo =TRUE, results='markup'}
sample_annot=read.table("data/sample_annotation.txt", head=T)
head(sample_annot)
groups=list(control=subset(sample_annot, group=="control"),
            treated=subset(sample_annot, group=="treated"))
groups$control
groups$treated

```

### Examine the count matrix

Since the count matrix is a numeric matrix,  we can perform standard numeric functions:

```{r operations, results='markup'}
# get the max of a data.matrix
max(count_data)
# get the min of a data.matrix
min(count_data)

# compute individual mean for all rows 
curr_rowMeans=rowMeans(count_data)
```

While these function are helpful,  there are important functions that do no compute for individual rows.  For example, standard deviation function **sd()** computes on a numeric vector. 

We know we can use **for** loop to compute sd row-by-row for all genes, but this will take a long time because sd operation is performed in a linear fashion.  Let's just try for the first 50 genes

```{r standard_deviation, results="markup"}
sd_by_genes=vector(mode="numeric")
for( gene in rownames(count_data)[1:50]){
   curr_sd=sd(count_data[gene,])
   sd_by_genes=c(curr_sd, sd_by_genes)
}
names(sd_by_genes)= rownames(count_data)[1:50]
```
This code will run for a long time on transcriptomic data.  The R function **apply()** applies a given function to different portions of a data.frame simultaneously to shorten the run time.

```
# the second input is called MARGIN, for a matrix 1 indicates rows, 2 indicates columns, c(1, 2) indicates rows and columns.
apply(input data.frame, 1 or 2, function_name )
```

```{R apply, results="markup"}
# the second input is called MARGIN, for a matrix 1 indicates rows, 2 indicates columns, c(1, 2) indicates rows and columns.
sd_by_genes=apply(count_data, 1, sd)
head(sd_by_genes)

```

### Use function to package

Often time we want to explore incoming data and save the exploratory results in a single variable as a list for comparison between data sets. We can put all above steps into **function** to make the procedure more reusable:

**function** takes the form of:
```
some_function_name<-function( arglist ) {
  operations
  return(value)
}
```
Let's try a simple example:

```{R simple_add, results="markup"}
add<-function(a,b){
  x=a+b
  return(x)
}

add(1,2)
add(3,4)
add(5,7)
```

Use **function** to compute max, min, row means, and row sd and package these information together with input data. We will do the followings in the function:
- compute max
- compute min
- comput mean expression for all genes
- comput sd of gene expression for all genes
- genereate a matrix with mean and sd annotation
- sort genes by standard deviation
- store everything in a list
- return the list

```{r basic_count_matrix_stat, results="markup"}
basic_rnaseq_stat<-function(input_data){
    exprs=input_data          # save input_data in a local variable
    exprs=count_data
    curr_max=max(exprs)       # compute max    
    curr_min=min(exprs)       # compute min
    gene_means=rowMeans(exprs)             # compute gene expression means
    gene_sd=apply(exprs, 1, sd)            # compute gene expression standard deviation  
    stat_df=merge(exprs, gene_means, by.x=0, by.y=0)         # merge means with expression matrix
    colnames(stat_df)[ncol(stat_df)]="mean"                  # change column name  
    stat_df=merge(stat_df, gene_sd, by.x=1, by.y=0)          # merge sd with expression matrix   
    colnames(stat_df)[ncol(stat_df)]="sd"                    # change column name 
    colnames(stat_df)[1]="gene"                              # change column name
    stat_df=stat_df[order(stat_df$sd, decreasing=TRUE),]     # sort the table based on Standard Deviation
    return(list(max=curr_max, min=curr_min, annot_matrix=stat_df))
}

exploratory_results=basic_rnaseq_stat(count_data)
names(exploratory_results)

# select top 20 most variable genes
top20_variable_genes=head(exploratory_results$annot_matrix, n=20)

```




## Principle component analysis
Once the matrix is loaded and processed with basic stat,  we can perform principle component analysis to visualize sample distribution.  
We first group samples into two vectors/variables: CONTROL and TREATED

```{r , echo=TRUE}
CONTROL=groups$control$sample
TREATED=groups$treated$sample
```

Calculate PCA use 'prcomp' command.  Since prcomp compute PCA by the rows, we will need to tranpose the count matrix such that the samples are represented by rows and genes by the columns.

```{r pca, echo=TRUE}
pca=prcomp(t(count_data))        # compute PCA on transposed count_data

plot(pca$x[,1], pca$x[,2], xlab='pc1', ylab='pc2')
points(pca$x[CONTROL,1], pca$x[CONTROL,2],
       col="red", pch=18)
points(pca$x[TREATED,1], pca$x[TREATED,2],
       col="blue", pch=18)
```

## Hierarchical Clustering

We first calculate similarity matrix using Euclidean distance.  The matrix is computed by 'dist' function. 'hclust' function computes hierarhical cluster tree and save it into an variable object 'hctree' that can be ploted by using generic 'plot' function.

```{r, echo=TRUE}
# compute similarity matrix on the transposed count_data matrix using Euclidean distance
dist_matrix=dist(t(count_data), method="euclidean")
# compute hierarchical clustering tree and save it into variable called 'hctree'
hctree=hclust(dist_matrix)
# plot hctree
plot(hctree)
```


## 
